### PR TITLE
asyncio for make-stats

### PR DESCRIPTION
**Pull request by mhauru**
Runs the make-stats queries in parallel.

### CODE REVIEW

File: sqlsynthgen/main.py [hunk 1]
```diff
@@ -167,7 +168,9 @@ def make_stats(
         typer.echo("Missing source database connection details.", err=True)
         sys.exit(1)
 
-    src_stats = make_src_stats(src_dsn, config)
+    src_stats = asyncio.get_event_loop().run_until_complete(
```
Comment (Iain-S) [inline sqlsynthgen/main.py:171]: If this change was purely for SSG then it sounds like `.run()` would be simplest but I guess we're calling `make_stats` from TAPAS so there might already be a loop running?

```suggestion
    src_stats = asyncio.get_event_loop().run_until_complete(
```

File: sqlsynthgen/make.py [hunk 2]
```diff
@@ -412,39 +406,45 @@ def make_src_stats(
...
     if use_smartnoise_sql:
+        logging.warning(
+            "SmartNoiseSQL does not support asyncio, so make-stats queries"
+            " will be run sequentially."
+        )
         dp_config = config.get("smartnoise-sql", {})
         snsql_metadata = {"": dp_config}
 
-        def execute_query(conn: Any, stat_data: Dict[str, Any]) -> Any:
+        async def execute_query(engine: Any, stat_data: Dict[str, Any]) -> Any:
```
Comment (Iain-S) [inline sqlsynthgen/make.py:422]: Suggest func comment and param rename:

```suggestion
        async def execute_query(engine: Any, query_block: Dict[str, Any]) -> Any:
            """Execute query using a synchronous SQLAlchemy engine."""
```

File: sqlsynthgen/make.py [hunk 3]
```diff
@@ -412,39 +406,45 @@ def make_src_stats(
...
+                )
+                private_result = reader.execute(stat_data["query"])
             # The first entry in the list names the columns, skip that.
             return private_result[1:]
 
     else:
 
-        def execute_query(conn: Any, stat_data: Dict[str, Any]) -> Any:
-            result = conn.execute(text(stat_data["query"])).fetchall()
+        async def execute_query(engine: Any, stat_data: Dict[str, Any]) -> Any:
```
Comment (Iain-S) [inline sqlsynthgen/make.py:439]: Suggest func comment and param rename:

```suggestion
        async def execute_query(engine: Any, query_block: Dict[str, Any]) -> Any:
            """Execute query using an asynchronous SQLAlchemy engine."""
```

File: sqlsynthgen/make.py [hunk 4]
```diff
@@ -412,39 +406,45 @@ def make_src_stats(
...
-    with engine.connect() as conn:
-        src_stats = {
-            stat_data["name"]: execute_query(conn, stat_data)
-            for stat_data in config.get("src-stats", [])
-        }
+    queries = config.get("src-stats", [])
+    results = await asyncio.gather(
+        *[execute_query(engine, stat_data) for stat_data in queries]
+    )
+    src_stats = {query["name"]: result for query, result in zip(queries, results)}
```
Comment (Iain-S) [inline sqlsynthgen/make.py:449]: Took me a while to work out what `stat_data` was. Is `query_block` any more intuitive?

```suggestion
    query_blocks = config.get("src-stats", [])
    results = await asyncio.gather(
        *[execute_query(engine, query_block) for query_block in query_blocks]
    )
    src_stats = {query_block["name"]: result for query_block, result in zip(query_blocks, results)}
```

File: sqlsynthgen/utils.py [hunk 5]
```diff
@@ -51,13 +52,24 @@ def download_table(table: Any, engine: Any, yaml_file_name: str) -> None:
...
+        async_dsn = postgres_dsn.replace("postgresql://", "postgresql+asyncpg://")
+        engine = create_async_engine(async_dsn, **kwargs)
+    else:
+        engine = create_engine(postgres_dsn, **kwargs)
+
+    if schema_name is not None:
+
+        @event.listens_for(engine, "connect", insert=True)
+        def connect(dbapi_connection: Any, _: Any) -> None:
+            set_search_path(dbapi_connection, schema_name)  # type: ignore
```
Comment (Iain-S) [inline sqlsynthgen/utils.py:72]: The schema event breaks if you try to attach it to an async engine.

```suggestion
        engine = create_async_engine(async_dsn, **kwargs)
        event_engine = engine.sync_engine
    else:
        engine = create_engine(postgres_dsn, **kwargs)
        event_engine = engine
        
    if schema_name is not None:

        @event.listens_for(event_engine, "connect", insert=True)
        def connect(dbapi_connection: Any, _: Any) -> None:
            set_search_path(dbapi_connection, schema_name)  # type: ignore
```

Comment (Iain-S) [reply sqlsynthgen/utils.py:72]: See #101

File: sqlsynthgen/main.py (hunk 1 above)
Comment (mhauru) [reply sqlsynthgen/main.py:171]: Yeaaah, I think that was my logic here. I'm not super sure of my understanding of how things like multiple event loops work though.

**Review by Iain-S — APPROVED** [review]
LGTM

**Pull request merged**